### ISSUE TITLE
Build has succeeded but webapp crashed H10 error code

### ISSUE THREAD

**Issue by lincredibleJC** [AdmitHub/meteor-buildpack-horse#177]
##### Type of problem
 - Runtime error - Build has succeeded but webapp crashed 

##### Summary
I've been looking at the error messages for about half a day now and also tried looking through the Heroku documentation to see how to fix the error but it still crashed.

I know that when cloning to desktop and using meteor there, there's this error that requires me to
meteor npm install --save babel-runtime
before I can serve meteor.
I'm not sure whether it's part of the reason that it might not be running here.
It definitely works on localhost though.

2017-07-04T11:33:29.551005+00:00 app[web.1]: Error: The babel-runtime npm package could not be found in your node_modules
2017-07-04T11:33:29.551006+00:00 app[web.1]:
2017-07-04T11:33:29.551005+00:00 app[web.1]: directory. Please run the following command to install it:
2017-07-04T11:33:29.551006+00:00 app[web.1]:   meteor npm install --save babel-runtime

Also, I recognise that the error seems to be from a crashed web dyno or a boot timeout on the web dyno, but I can't find any solution to solve the problem.

2017-07-04T11:43:23.588123+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=nushortcuts.herokuapp.com request_id=720c77f5-b9b0-42fe-acb4-afde8d917efb fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https
2017-07-04T11:43:26.141915+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=nushortcuts.herokuapp.com request_id=35697dd6-e1f6-4d12-b84d-4bd123e53f87 fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https

##### Contents of `.meteor/release`
METEOR@1.5

##### Buildpack version
https://github.com/AdmitHub/meteor-buildpack-horse.git

##### Logs
If the build was successful and this is a runtime problem, please paste any
relevant logs from `heroku logs`.
2017-07-04T11:28:13.276931+00:00 app[api]: Enable Logplex by user lin_jiachao@hotmail.com
2017-07-04T11:28:25.060192+00:00 app[api]: Attach MONGODB (@ref:mongolab-metric-41142) by user lin_jiachao@hotmail.com
2017-07-04T11:28:25.060192+00:00 app[api]: Release v3 created by user lin_jiachao@hotmail.com
2017-07-04T11:29:20.224266+00:00 app[api]: Set MONGO_URL config vars by user lin_jiachao@hotmail.com
2017-07-04T11:29:20.224266+00:00 app[api]: Release v4 created by user lin_jiachao@hotmail.com
2017-07-04T11:30:04.657860+00:00 app[api]: Set ROOT_URL config vars by user lin_jiachao@hotmail.com
2017-07-04T11:30:04.657860+00:00 app[api]: Release v5 created by user lin_jiachao@hotmail.com
2017-07-04T11:30:58.000000+00:00 app[api]: Build started by user lin_jiachao@hotmail.com
2017-07-04T11:33:22.786219+00:00 app[api]: Scaled to web@1:Free by user lin_jiachao@hotmail.com
2017-07-04T11:33:22.774067+00:00 app[api]: Release v6 created by user lin_jiachao@hotmail.com
2017-07-04T11:33:22.774067+00:00 app[api]: Deploy 4613a12e by user lin_jiachao@hotmail.com
2017-07-04T11:30:58.000000+00:00 app[api]: Build succeeded
2017-07-04T11:33:26.801900+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-04T11:33:29.550990+00:00 app[web.1]: /app/.meteor/heroku_build/app/programs/server/boot.js:391
2017-07-04T11:33:29.551003+00:00 app[web.1]: }).run();
2017-07-04T11:33:29.551003+00:00 app[web.1]:    ^
2017-07-04T11:33:29.551004+00:00 app[web.1]:
2017-07-04T11:33:29.551005+00:00 app[web.1]: Error: The babel-runtime npm package could not be found in your node_modules
2017-07-04T11:33:29.551006+00:00 app[web.1]:
2017-07-04T11:33:29.551005+00:00 app[web.1]: directory. Please run the following command to install it:
2017-07-04T11:33:29.551006+00:00 app[web.1]:   meteor npm install --save babel-runtime
2017-07-04T11:33:29.551007+00:00 app[web.1]:
2017-07-04T11:33:29.551008+00:00 app[web.1]:     at meteorInstall.node_modules.meteor.babel-runtime.babel-runtime.js (packages/babel-runtime.js:36:9)
2017-07-04T11:33:29.551008+00:00 app[web.1]:     at fileEvaluate (packages/modules-runtime.js:333:9)
2017-07-04T11:33:29.551008+00:00 app[web.1]:     at require (packages/modules-runtime.js:228:16)
2017-07-04T11:33:29.551009+00:00 app[web.1]:     at packages/babel-runtime.js:143:15
2017-07-04T11:33:29.551009+00:00 app[web.1]:     at packages/babel-runtime.js:154:3
2017-07-04T11:33:29.551011+00:00 app[web.1]:     at Function._.each._.forEach (/app/.meteor/heroku_build/app/programs/server/node_modules/underscore/underscore.js:79:11)
2017-07-04T11:33:29.551010+00:00 app[web.1]:     at /app/.meteor/heroku_build/app/programs/server/boot.js:338:34
2017-07-04T11:33:29.551010+00:00 app[web.1]:     at Array.forEach (native)
2017-07-04T11:33:29.551012+00:00 app[web.1]:     at /app/.meteor/heroku_build/app/programs/server/boot.js:387:5
2017-07-04T11:33:29.551011+00:00 app[web.1]:     at /app/.meteor/heroku_build/app/programs/server/boot.js:158:5
2017-07-04T11:33:29.619972+00:00 heroku[web.1]: Process exited with status 1
2017-07-04T11:33:29.631177+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-04T11:33:29.632205+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-04T11:33:33.985815+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-04T11:33:37.458915+00:00 app[web.1]: /app/.meteor/heroku_build/app/programs/server/boot.js:391
2017-07-04T11:33:37.458928+00:00 app[web.1]: }).run();
2017-07-04T11:33:37.458929+00:00 app[web.1]:    ^
2017-07-04T11:33:37.458929+00:00 app[web.1]:
2017-07-04T11:33:37.458930+00:00 app[web.1]: Error: The babel-runtime npm package could not be found in your node_modules
2017-07-04T11:33:37.458931+00:00 app[web.1]: directory. Please run the following command to install it:
2017-07-04T11:33:37.458931+00:00 app[web.1]:
2017-07-04T11:33:37.458933+00:00 app[web.1]:
2017-07-04T11:33:37.458932+00:00 app[web.1]:   meteor npm install --save babel-runtime
2017-07-04T11:33:37.458934+00:00 app[web.1]:     at meteorInstall.node_modules.meteor.babel-runtime.babel-runtime.js (packages/babel-runtime.js:36:9)
2017-07-04T11:33:37.458934+00:00 app[web.1]:     at fileEvaluate (packages/modules-runtime.js:333:9)
2017-07-04T11:33:37.458935+00:00 app[web.1]:     at require (packages/modules-runtime.js:228:16)
2017-07-04T11:33:37.458935+00:00 app[web.1]:     at packages/babel-runtime.js:143:15
2017-07-04T11:33:37.458936+00:00 app[web.1]:     at packages/babel-runtime.js:154:3
2017-07-04T11:33:37.458937+00:00 app[web.1]:     at /app/.meteor/heroku_build/app/programs/server/boot.js:338:34
2017-07-04T11:33:37.458937+00:00 app[web.1]:     at Array.forEach (native)
2017-07-04T11:33:37.458938+00:00 app[web.1]:     at Function._.each._.forEach (/app/.meteor/heroku_build/app/programs/server/node_modules/underscore/underscore.js:79:11)
2017-07-04T11:33:37.458939+00:00 app[web.1]:     at /app/.meteor/heroku_build/app/programs/server/boot.js:387:5
2017-07-04T11:33:37.458939+00:00 app[web.1]:     at /app/.meteor/heroku_build/app/programs/server/boot.js:158:5
2017-07-04T11:33:37.580268+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-04T11:33:37.563430+00:00 heroku[web.1]: Process exited with status 1
2017-07-04T11:34:10.878987+00:00 app[api]: Set PATH config vars by user lin_jiachao@hotmail.com
2017-07-04T11:34:10.878987+00:00 app[api]: Release v7 created by user lin_jiachao@hotmail.com
2017-07-04T11:34:11.942636+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-04T11:34:14.430755+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-04T11:34:15.977668+00:00 app[web.1]: Error: No such file or directory
2017-07-04T11:34:15.977925+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-04T11:34:16.045811+00:00 heroku[web.1]: Process exited with status 22
2017-07-04T11:34:16.060938+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-04T11:38:31.000000+00:00 app[api]: Build started by user lin_jiachao@hotmail.com
2017-07-04T11:40:27.551646+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-04T11:40:27.265551+00:00 app[api]: Deploy 4613a12e by user lin_jiachao@hotmail.com
2017-07-04T11:40:27.265551+00:00 app[api]: Release v8 created by user lin_jiachao@hotmail.com
2017-07-04T11:38:31.000000+00:00 app[api]: Build succeeded
2017-07-04T11:40:30.537847+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-04T11:40:32.505611+00:00 heroku[web.1]: Process exited with status 22
2017-07-04T11:40:32.420429+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-04T11:40:32.411178+00:00 app[web.1]: Error: No such file or directory
2017-07-04T11:40:32.516717+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-04T11:43:23.588123+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=nushortcuts.herokuapp.com request_id=720c77f5-b9b0-42fe-acb4-afde8d917efb fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https
2017-07-04T11:43:26.141915+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=nushortcuts.herokuapp.com request_id=35697dd6-e1f6-4d12-b84d-4bd123e53f87 fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https
2017-07-04T11:44:29.431583+00:00 app[api]: Set BUILDPACK_VERBOSE config vars by user lin_jiachao@hotmail.com
2017-07-04T11:44:29.431583+00:00 app[api]: Release v9 created by user lin_jiachao@hotmail.com
2017-07-04T11:44:29.754247+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-04T11:44:32.387596+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-04T11:44:34.127237+00:00 heroku[web.1]: Process exited with status 22
2017-07-04T11:44:34.073572+00:00 app[web.1]: Error: No such file or directory
2017-07-04T11:44:34.073877+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-04T11:44:34.140376+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-04T11:44:34.141525+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-04T11:44:36.963255+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-04T11:44:39.484069+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-04T11:44:39.393762+00:00 app[web.1]: Error: No such file or directory
2017-07-04T11:44:39.394094+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-04T11:44:39.471778+00:00 heroku[web.1]: Process exited with status 22
2017-07-04T11:45:06.000000+00:00 app[api]: Build started by user lin_jiachao@hotmail.com
2017-07-04T11:46:51.029888+00:00 app[api]: Deploy 4613a12e by user lin_jiachao@hotmail.com
2017-07-04T11:46:51.029888+00:00 app[api]: Release v10 created by user lin_jiachao@hotmail.com
2017-07-04T11:46:51.941161+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-04T11:45:06.000000+00:00 app[api]: Build succeeded
2017-07-04T11:46:54.385875+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-04T11:46:55.882954+00:00 app[web.1]: Error: No such file or directory
2017-07-04T11:46:55.883187+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-04T11:46:55.939061+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-04T11:46:55.926962+00:00 heroku[web.1]: Process exited with status 22

##### Deployment target
Deploying to Heroku

**Comment by yourcelf** [comment]
Looks like the operative error is `The babel-runtime npm package could not be found in your node_modules`.  Did you install that library and save it to package.json?

See: https://stackoverflow.com/questions/40499685/error-with-babel-runtime-while-trying-to-serve-a-meteor-app

**Comment by lincredibleJC** [comment]
Yes, I've just added a package.json and added the babel-runtime and the other dependencies needed for it to run. The app still runs on localhost but still does not work on heroku.

this is the heroku logs now:

2017-07-14T07:57:55.460480+00:00 app[api]: Release v17 created by user lin_jiachao@hotmail.com
2017-07-14T07:57:55.460480+00:00 app[api]: Deploy 61195e8d by user lin_jiachao@hotmail.com
2017-07-14T07:57:55.817119+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-14T07:55:44.000000+00:00 app[api]: Build succeeded
2017-07-14T07:57:59.503018+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-14T07:58:01.176756+00:00 app[web.1]: Error: No such file or directory
2017-07-14T07:58:01.177049+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-14T07:58:01.254613+00:00 heroku[web.1]: Process exited with status 22
2017-07-14T07:58:01.271252+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-14T08:05:07.698919+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=nushortcuts.herokuapp.com request_id=d9af912f-be71-4b99-83be-553df2b30922 fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https
2017-07-14T08:05:10.160956+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=nushortcuts.herokuapp.com request_id=5ef964ac-aa23-43b1-840c-bf6898cd81be fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https
2017-07-14T08:14:11.725100+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-14T08:14:15.499431+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-14T08:14:17.343038+00:00 heroku[web.1]: Process exited with status 22
2017-07-14T08:14:17.282713+00:00 app[web.1]: Error: No such file or directory
2017-07-14T08:14:17.282998+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-14T08:14:17.354437+00:00 heroku[web.1]: State changed from starting to crashed

**Comment by yourcelf** [comment]
The operative error seems to be `Error: No such file or directory`. Are all the files required by your app checked into git?

**Comment by matadur** [comment]
Do you have a `main.js` that's not checked in?

Does it run locally when you run `meteor --production`?

**Comment by lincredibleJC** [comment]
@matadur yes it runs on meteor --production

Sorry but I don't really understand what I'm missing here. I'm probably understanding meteor's file structure wrongly.

It seems that main.js which refers to the entry point for the client and server is not checked in. But in my meteor client folder, my entry point is named something different: layout.html.
So I changed that to main.html and added a main.js which was previously not needed. Then I created a server folder and added in another main.js with a meteor.startup(). I changed main in package.json to main.js and I deployed again.

But after doing that, I'm still getting the no such file directory error.
Seems lilke there is something that I'm definitely not understanding here but I just cant figure out what.


2017-07-15T05:30:08.060772+00:00 app[web.1]: Error: No such file or directory
2017-07-15T05:30:08.185322+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-15T05:30:08.186693+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-15T05:30:08.168928+00:00 heroku[web.1]: Process exited with status 126
2017-07-15T05:30:13.220811+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-15T05:30:15.496780+00:00 app[web.1]: Error: No such file or directory
2017-07-15T05:30:15.497381+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-15T05:30:15.601079+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-15T05:30:15.585870+00:00 heroku[web.1]: Process exited with status 22
2017-07-15T05:30:17.566955+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=nushortcuts.herokuapp.com request_id=cce08c72-e67e-4ad0-8f04-a58fd0d34147 fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https
2017-07-15T05:30:19.481725+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=nushortcuts.herokuapp.com request_id=fb489689-62f1-4126-952f-359341d52f09 fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https

**Comment by matadur** [comment]
Yeah it shouldn't be needed - I think the main.js it's looking for is actually the one generated by the Meteor build. I can't tell from the log if it can't find that one, or if it's actually starting and then immediately can't find something else (that's not in the log).

Have you tried clearing the buildpack cache? I wonder if it's using a previously bad build from your earlier errors.

BUILDPACK_CLEAR_CACHE: This buildpack stores the meteor installation in the CACHE_DIR to speed up subsequent builds. Set BUILDPACK_CLEAR_CACHE=1 to clear this cache on startup and after build is done.

**Comment by lincredibleJC** [comment]
Just tried that. Same error code. No luck still.
I just tried deploying a foobar example heroku app and it works.
I think I might try to deploy an empty template and incrementally rebuild the whole app in a separate repo. That might show which step caused the error.


2017-07-15T17:29:58.115973+00:00 app[api]: Deploy e0f19c4b by user lin_jiachao@hotmail.com
2017-07-15T17:29:58.115973+00:00 app[api]: Release v21 created by user lin_jiachao@hotmail.com
2017-07-15T17:27:50.000000+00:00 app[api]: Build succeeded
2017-07-15T17:29:59.880714+00:00 heroku[web.1]: State changed from crashed to starting
2017-07-15T17:30:03.835463+00:00 heroku[web.1]: Starting process with command `.meteor/heroku_build/bin/node $NODEJS_PARAMS .meteor/heroku_build/app/main.js`
2017-07-15T17:30:05.561572+00:00 app[web.1]: Error: No such file or directory
2017-07-15T17:30:05.562005+00:00 app[web.1]: Error waiting for process to terminate: No child processes
2017-07-15T17:30:05.621822+00:00 heroku[web.1]: Process exited with status 22
2017-07-15T17:30:05.631267+00:00 heroku[web.1]: State changed from starting to crashed
2017-07-15T17:34:55.850458+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=nushortcuts.herokuapp.com request_id=e2070c3e-0c08-42f2-a399-0d0d6cce7ee0 fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https
2017-07-15T17:34:58.104441+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/favicon.ico" host=nushortcuts.herokuapp.com request_id=916e1511-29e3-4b60-8f53-b44ecfbae826 fwd="116.14.32.199" dyno= connect= service= status=503 bytes= protocol=https

**Comment by lincredibleJC** [comment]
Hi, just an update. I've just solved the problem. Apparently, while deploying the example foobar app, I realised that if it works without PATH=$HOME/.meteor then the app no longer crashes. So all is good here.

![image](https://user-images.githubusercontent.com/22949932/28245182-74e7bf30-6a33-11e7-9d78-8a87805dc39b.png)

However, I'm still a little confused about what's wrong with the PATH config.
Is HOME supposed to be a placeholder? If it is, I've already tried inputting the correct directory name, but that crashes the app again.
So I'm a little confused here. But for now, I've just removed it and it works now.

Thanks for the help in debugging this problem!!
